pkg/kubernetes/artefact_downloader.go

```diff
@@ -0,0 +1,128 @@
+
+	rke2CalicoImages = "rke2-images-calico.linux-%s.tar.zst"
+	rke2CanalImages  = "rke2-images-canal.linux-%s.tar.zst"
+	rke2CiliumImages = "rke2-images-cilium.linux-%s.tar.zst"
+	rke2MultusImages = "rke2-images-multus.linux-%s.tar.zst"
+
+	rke2VSphereImages = "rke2-images-vsphere.linux-%s.tar.zst"
+)
+
+type ArtefactDownloader struct{}
+
+func (d ArtefactDownloader) DownloadArtefacts(kubernetes image.Kubernetes, arch image.Arch, destinationPath string) error {
+	if !strings.Contains(kubernetes.Version, image.KubernetesDistroRKE2) {
+		return fmt.Errorf("kubernetes version '%s' is not supported", kubernetes.Version)
+	}
+
+	path := filepath.Join(destinationPath, kubernetesDir)
+
+	if err := os.Mkdir(path, os.ModePerm); err != nil {
+		return fmt.Errorf("creating kubernetes dir: %w", err)
+	}
+
+	artefacts, err := gatherArtefacts(kubernetes, arch)
+	if err != nil {
+		return fmt.Errorf("gathering RKE2 artefacts: %w", err)
+	}
+
+	if err = downloadArtefacts(artefacts, rke2ReleaseURL, kubernetes.Version, destinationPath); err != nil {
+		return fmt.Errorf("downloading RKE2 artefacts: %w", err)
+	}
+
+	return nil
+}
+
+func gatherArtefacts(kubernetes image.Kubernetes, arch image.Arch) ([]string, error) {
+	if arch == image.ArchTypeARM {
+		log.Audit("WARNING: RKE2 support for aarch64 platforms is limited and experimental")
+	}
+
+	artefactArch := arch.Short()
+
+	var artefacts []string
+
+	artefacts = append(artefacts,
+		fmt.Sprintf(rke2Binary, artefactArch),
+		fmt.Sprintf(rke2CoreImages, artefactArch),
+		fmt.Sprintf(rke2Checksums, artefactArch))
+
+	switch kubernetes.CNI {
+	case "":
+		return nil, fmt.Errorf("CNI not specified")
+	case image.CNITypeNone:
+	case image.CNITypeCanal:
+		artefacts = append(artefacts, fmt.Sprintf(rke2CanalImages, artefactArch))
+	case image.CNITypeCalico:
+		if arch == image.ArchTypeARM {
+			return nil, fmt.Errorf("calico is not supported on %s platforms", arch)
+		}
+		artefacts = append(artefacts, fmt.Sprintf(rke2CalicoImages, artefactArch))
+	case image.CNITypeCilium:
+		if arch == image.ArchTypeARM {
+			return nil, fmt.Errorf("cilium is not supported on %s platforms", arch)
+		}
+		artefacts = append(artefacts, fmt.Sprintf(rke2CiliumImages, artefactArch))
+	default:
+		return nil, fmt.Errorf("unsupported CNI: %s", kubernetes.CNI)
+	}
+
+	if kubernetes.MultusEnabled {
+		if arch == image.ArchTypeARM {
+			return nil, fmt.Errorf("multus is not supported on %s platforms", arch)
+		}
+		artefacts = append(artefacts, fmt.Sprintf(rke2MultusImages, artefactArch))
+	}
+
+	if kubernetes.VSphereEnabled {
+		if arch == image.ArchTypeARM {
+			return nil, fmt.Errorf("vSphere is not supported on %s platforms", arch)
+		}
+		artefacts = append(artefacts, fmt.Sprintf(rke2VSphereImages, artefactArch))
+	}
+
+	return artefacts, nil
+}
+
+func downloadArtefacts(artefacts []string, releaseURL, version, destinationPath string) error {
+	errGroup, ctx := errgroup.WithContext(context.Background())
```

Comment:
A timeout is a bad idea, especially because we have no idea what sort of network connections our customers will have available to them, especially if they're behind a proxy. We may want to do some kind of calculation at a later date where we get the filesizes before we download and ask them to confirm that they want to download these files (with an unattended override for automation, of course). It also makes me question whether we would also want to allow the customer to side-load these, and where a download is skipped if these files are provided manually? There's a risk with this, but I can imagine a situation where the customer wouldn't want EIB to download them.

Reply:
We've talked about the second point in the sense of caching between builds; no need to eat the k8s download for repeated builds. It's too much for 3.0 and we'll need to revisit later.

As for the timeout, I agree it's difficult to find a good value, but at the same time I don't want it to appear as if EIB is frozen. That's probably fine for now until we get more users and see how it shakes out for them.

Reply:
We'll fix the UX bit by introducing the progress bar we briefly discussed yesterday.